keps/sig-api-machinery/2211-strip-managed-fields-server-side/README.md

```diff
@@ -0,0 +1,426 @@
+
+The canonical place for the latest set of instructions (and the likely source
+of this file) is [here](/keps/NNNN-kep-template/README.md).
+
+**Note:** Any PRs to move a KEP to `implementable`, or significant changes once
+it is marked `implementable`, must be approved by each of the KEP approvers.
+If none of those approvers are still appropriate, then changes to that list
+should be approved by the remaining approvers and/or the owning SIG (or
+SIG Architecture for cross-cutting KEPs).
+-->
+
+# KEP-2211: Strip ManagedFields on the Server-side
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Server-side changes](#server-side-changes)
+  - [Client-side changes](#client-side-changes)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Server-side](#server-side)
+  - [Client-side](#client-side)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+ManagedFields is used to keep track of who has changed a field of an object, and Server-side Apply relies on such information.
+While the managedFields is beneficial to resolving conflicts on updates, it is a bit verbose to many users because this field
+could be quite large and generally users are not interested in this information.
+
+This KEP is proposing stripping the managedFields on the server-side when users request objects to make the response less verbose.
```

Comment:
Not only when user "requests objects", since pretty much all operations return the final object.

Reply:
Hopefully we are talking about omitting them from the response and not stripping them from the stored object.

Reply:
Why not omitting it from stored object as well ? If a cluster operator has made a conscious decision,  it would save a lot of storage for him.

Reply:
> Hopefully we are talking about omitting them from the response and not stripping them from the stored object.

Yes, the original object saved in etcd is left intact.

> Why not omitting it from stored object as well 

The Server-side Apply feature needs the managedFields.

Reply:
> If a cluster operator has made a conscious decision, it would save a lot of storage for him.

It's unclear if the trade-offs they're making would be all that conscious.

I'm happy to discuss if there is a specific use-case that you have in mind, but the "Accept" header is not where we would want to strip managedFields from storage.

Reply:
> Why not omitting it from stored object as well ?

1. it loses valuable context about the object that other aspects of the system need
2. this would not be the right mechanism for that at all, we'd need to make a setting otherwise it would come back
3. this is primarily about read operations, although it will also work on write operations

Reply:
+1 to Daniel. Managed field is crucial to automatic config mgmt and a depency for system needs. This proposal should be a minimal UI fix IMO.

A off-topic idea is we may want to bring up some real-world cases how this tech could be properly used by the platform team in server-side apply doc. The main use case on our side is k-o-k configuration mgmt scenario, but it should be more.

Reply:
Sounds good thanks. I agree a different setting is needed. I think until this becomes widespread in usage, the default should be ignored in returns and not stored as well. This is ahead of its time , but definitely a really nice concept.